File: python/querygraph/codata.py
from __future__ import annotations

import json
from dataclasses import dataclass
from urllib.parse import urlencode
from urllib.request import urlopen


@dataclass(frozen=True)
class StoredPayload:
    url: str | None = None
    timestamp: str | None = None
    title: str | None = None
    is_rdf: bool | None = None


@dataclass(frozen=True)
class AnchoredDid:
    did: str
    doc: dict | None = None
    stored_payload: StoredPayload | None = None


class CodataOdrlClient:
    def __init__(self, base_url: str = "https://odrl.dev.codata.org") -> None:
        self.base_url = base_url.rstrip("/")

    def create_did_from_url(self, url: str) -> AnchoredDid:
        query = urlencode({"url": url})
        with urlopen(f"{self.base_url}/api/did/create_from_url?{query}") as response:
            payload = json.loads(response.read().decode())

        stored_payload = payload.get("stored_payload")
        return AnchoredDid(
            did=payload["did"],
            doc=payload.get("doc"),
            stored_payload=StoredPayload(**stored_payload) if stored_payload else None,
        )

File: src/codata.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct StoredPayload {
    pub url: Option<String>,
    pub timestamp: Option<String>,
    pub title: Option<String>,
    pub is_rdf: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct AnchoredDid {
    pub did: String,
    pub doc: Option<Value>,
    pub stored_payload: Option<StoredPayload>,
}

#[derive(Debug, Clone)]
pub struct CodataOdrlClient {
    base_url: String,
}

impl Default for CodataOdrlClient {
    fn default() -> Self {
        Self::new("https://odrl.dev.codata.org")
    }
}

impl CodataOdrlClient {
    pub fn new(base_url: impl Into<String>) -> Self {
        Self {
            base_url: base_url.into().trim_end_matches('/').to_string(),
        }
    }

    pub fn create_did_from_url(&self, url: &str) -> Result<AnchoredDid, reqwest::Error> {
        reqwest::blocking::Client::new()
            .get(format!("{}/api/did/create_from_url", self.base_url))
            .query(&[("url", url)])
            .send()?
            .error_for_status()?
            .json()
    }
}
